### ISSUE TITLE
Issue on training step

### ISSUE THREAD

**Issue by Azazel17** [jkjung-avt/hand-detection-tutorial#12]
Sir. Excellent work i have the following error when running the training step:

2019-04-07 00:03:56.985039: E tensorflow/stream_executor/cuda/cuda_dnn.cc:396] Loaded runtime CuDNN library: 7401 (compatibility version 7400) but source was compiled with 7005 (compatibility version 7000).  If using a binary install, upgrade your CuDNN library to match.  If building from sources, make sure the library loaded at runtime matches a compatible version specified during compile configuration.
2019-04-07 00:03:56.986207: F tensorflow/core/kernels/conv_ops.cc:712] Check failed: stream->parent()->GetConvolveAlgorithms( conv_parameters.ShouldIncludeWinogradNonfusedAlgo<T>(), &algorithms) 
./train.sh: línea 60: 17698 Abortado                (`core' generado) PYTHONPATH=`pwd`/models/research:`pwd`/models/research/slim python3 ./models/research/object_detection/model_main.py --pipeline_config_path=${PIPELINE_CONFIG_PATH} --model_dir=${MODEL_DIR} --num_train_steps=${NUM_TRAIN_STEPS} --sample_1_of_n_eval_samples=1 --alsologtostderr

Any help would be appreciated, by the way awesome work

**Comment by jkjung-avt** [comment]
Thanks for the compliment.

The error message you posted says that you have a newer version (7401) of [cuDNN](https://developer.nvidia.com/cudnn) library installed on your system while trying to run a tensorflow binary compiled against an older version (7000) of cuDNN.  And the program aborted due to this incompatibility.

Try a newer version of tensorflow, or build the tensorflow binary by yourself.

**Comment by Azazel17** [comment]
Thank you sr. for your help,( is already working by the way),  i have a question.  i'm trying to perform detection on relatively small images. But the detection is not so good. Is there anyway i can improve the detection based on your work? I attached some images that i'm trying to detect. Thanks in advance for your help 

![detection_output](https://user-images.githubusercontent.com/21000114/55695518-84c89500-597e-11e9-82ca-857963cb800b.jpg)
![detection_output](https://user-images.githubusercontent.com/21000114/55695543-a75aae00-597e-11e9-8bb0-7de7cef7ecb0.jpg)

**Comment by jkjung-avt** [comment]
All object detection models, including SSD, in general don't perform very well on detecting small objects.  For the pictures shown above, you could try to "crop only the center area of the images" so that the hands don't appear too small on the scaled-down (300x300) image, i.e. input to the SSD model.

Otherwise, you might try to set 'ssd_anchor_generator'->'min_scale' to a smaller value (which I already did) to force the SSD model to learn to detect smaller objects.

https://github.com/jkjung-avt/hand-detection-tutorial/blob/master/configs/ssd_mobilenet_v1_egohands.config#L34

In my example config, the 'min_scale' is set to 0.1, which corresponds to objects of 30 pixels wide/high on the 300x300 image.  Assuming the original image is 640x640, then the SSD model is probably only capable of detecting objects (hands) of at least 64 pixels in size (30 pixels / 300 * 640)...

**Comment by Azazel17** [comment]
Thank you for your help sr, i understand, is there any other model/network arquitecture that could improve results? also can i retrieve the coordinates of the box detectors? By the way i'm retrining modifyng the parameters to 0.1 because i was using mobilenet_V2 before 

Thanks in advance i really appreciate your help

**Comment by jkjung-avt** [comment]
1. Using a faster_rcnn model with high resolution (say, 1000x600) image input might help.  But it would run much slowlier.
2. Check out [detect_image.py line $59](https://github.com/jkjung-avt/hand-detection-tutorial/blob/master/detect_image.py#L59), bounding box coordinates are stored in the numpy array 'boxes'.

**Comment by Azazel17** [comment]
Really appreciate your help, i say it again your work is awesome it works better now using faster_rcnn_inception_v2_egohands. 

However the precission is low in some frames, how change the scale on this network? Can't find the min_scale parameter.  

I'm going to check the 'boxes' var. 

Again sr, thank you very much.

**Comment by jkjung-avt** [comment]
For 'faster_rcnn_inception_v2_egohands', I think the 'scales' ([reference](https://github.com/jkjung-avt/hand-detection-tutorial/blob/master/configs/faster_rcnn_inception_v2_egohands.config#L20)) in the anchor generator roughly correspond to how small/large the targeted objects are.  The smallest scale, 0.25, would roughly determine the smallest objects the faster_rcnn model could detect.

I probably have done some experiment (tried to set a smaller scale) with that, but I recall it did not work well.  One of the reasons might be the pre-trained weights (on coco dataset) did not learn to detect such smaller objects.

```
    first_stage_anchor_generator {
      grid_anchor_generator {
        scales: [0.25, 0.5, 1.0, 2.0]
        aspect_ratios: [0.5, 1.0, 2.0]
        height_stride: 16
        width_stride: 16
      }
    }
```

**Comment by Azazel17** [comment]
Sr. thanks again for your help in fact i use the following config 

`    first_stage_anchor_generator {
      grid_anchor_generator {
        scales: [0.075, 0.12, 0.25, 0.5, 1.0, 2.0]
        aspect_ratios: [0.12, 0.25, 0.5, 1.0, 2.0]
        height_stride: 16
        width_stride: 16
      }
    }`

And improves the classification on small images, even without having the background segmented. Also caoul you please provide me the pretrained models please? By the way when i print the 'boxes' var, i get the following output: 

`[[[0.5738091  0.10619719 0.65711313 0.21840563]
 [0.19802608 0.39142582 0.26067692 0.56693834]
  [0.20598033 0.4226063  0.36812422 0.8793116 ]
  ...
  [0.         0.         0.         0.        ]
  [0.         0.         0.         0.        ]
  [0.         0.         0.         0.        ]]]
`

How can i interpret this output? 

Thanks again four your help

**Comment by jkjung-avt** [comment]
1. Regarding the 'aspect_ratios', the original design of [0.5, 1.0, 2.0]  (corresponding 1:2, 1:1 and 2:1 aspect ratios of objects) should be fine.  I think you don't need to add the 0.12 and 0.25.  Those would just slow down the model without much benefit.

2. The 'boxes' variables should be interpreted together with 'classes' and 'scores':  (assuming the SSD model is designed to generate at maximum N detections)
    'boxes': Nx4 numpy array of bounding box coordinates.  The 4 numbers for each bounding box correspond to [x1, y1, x2, y2] values.
    'classes': Nx1 numpy array of detection class IDs.
    'scores': Nx1 numpy array of confidence score of the corresponding detection.  Each score value is between 0 and 1.

The [x1, y1, x2, y2] coordinates are normalized with respect to image width and height.  So take your example (1st detection in your 'boxes'): [0.5738091 0.10619719 0.65711313 0.21840563].  If the input image shape is 1280x720, then the absolute [x1, y1, x2, y2] should be [0.5738091*1280, 0.10619719*720, 0.65711313*1280, 0.21840563*720] = [734.5, 76.5, 841.1, 157.3].
